vault/src/main/java/org/opencadc/vault/metadata/DataNodeSizeSync.java

```diff
@@ -0,0 +1,225 @@
+    private static final long EVICT_AGE = 12 * ROUNDS;
+    
+    private static final long FAIL_SLEEP = 10 * ROUNDS; // slightly smaller than evict
+    
+    private final UUID instanceID = UUID.randomUUID();
+    private final HarvestStateDAO stateDAO;
+    private final ArtifactDAO artifactDAO;
+    private final Namespace artifactNamespace;
+    
+    private boolean offline = false;
+    
+    public DataNodeSizeSync(HarvestStateDAO stateDAO, ArtifactDAO artifactDAO, Namespace artifactNamespace) { 
+        this.stateDAO = stateDAO;
+        this.artifactDAO = artifactDAO;
+        this.artifactNamespace = artifactNamespace;
+        
+        // we need continuous timestamp updates to retain leader status, so only schedule maintenance
+        stateDAO.setUpdateBufferCount(0);
+        stateDAO.setMaintCount(9999); // every 1e4
+    }
+
+    public void setOffline(boolean offline) {
+        this.offline = offline;
+    }
+
+    @Override
+    public void run() {
+        String name = Artifact.class.getSimpleName();
+        URI resourceID = URI.create("jdbc:inventory");
+        try {
+            Thread.sleep(1 * ROUNDS); // delay startup a bit
+            
+            while (true) {
+                while (offline) {
+                    log.warn("disabled: sleep=" + LONG_SLEEP);
+                    Thread.sleep(LONG_SLEEP);
+                }
+
+                log.debug("check leader " + instanceID);
+                HarvestState state = stateDAO.get(name, resourceID);
+                log.debug("check leader " + instanceID + " found: " + state);
+                if (state.instanceID == null) {
+                    state.instanceID = instanceID;
+                    stateDAO.put(state);
+                    state = stateDAO.get(state.getID());
+                    log.debug("created: " + state);
+                }
+                
+                long t1 = System.currentTimeMillis();
+                BackgroundLogInfo logInfo = new BackgroundLogInfo(instanceID.toString());
+                logInfo.setOperation(DataNodeSizeWorker.class.getSimpleName());
+                logInfo.setSuccess(false);
+                
+                // determine leader
+                boolean leader = checkLeaderStatus(state);
+                logInfo.leader = leader;
+                logInfo.setLastModified(state.curLastModified);
+                long sleep = LONG_SLEEP;
+                if (leader) {
+                    log.debug("leader: " + state);
+                    boolean fail = false;
+                    log.info(logInfo.start());
+                    try {
+                        DataNodeSizeWorker worker = new DataNodeSizeWorker(stateDAO, state, artifactDAO, artifactNamespace);
+                        worker.run();
+                        logInfo.setLastModified(state.curLastModified);
+                        logInfo.processed = worker.getNumArtifactsProcessed();
+                        logInfo.setSuccess(true);
```

Comment:
Isn't this premature since it can be set back to `false` later?

Reply:
yes, technically correct. fixed